File: Algorithm/inflearn_cpp/myCode/90_라이언킹_심바_역량기출.py
n = int(input())
a = [list(map(int, input().split())) for _ in range(n)]
chk = [[False for _ in range(n)] for __ in range(n)]
size = 2
ate = 0
rabbits = 0
q = []

for i in range(n):
    for j in range(n):
        if a[i][j] == 9:
            q.append([i, j, 0])
            a[i][j] = 0
        elif a[i][j]:
            rabbits += 1
while q:
    r, c, time = q[0]
    chk[r][c] = True
    q = q[1:]
    for d in range(4):
        nr, nc = r + dr[d], c + dc[d]
        ntime = time + 1
        if nr < 0 or nc < 0 or nr >= n or nc >= n or chk[nr][nc]:
            continue
        if a[nr][nc] > 0 and a[nr][nc] < size:
            ate += 1
            rabbits -= 1
            a[nr][nc] = 0
            if ate == size:
                size += 1
                ate = 0
            chk = [[False for _ in range(n)] for __ in range(n)]
            q = []
            q.append([nr, nc, ntime])
            break
        elif a[nr][nc] <= size:
            q.append([nr, nc, ntime])
print(time)

File: Algorithm/inflearn_cpp/myCode/90_라이언킹_심바_역량기출.cpp
#include <stdio.h>
#include <queue>
using namespace std;
int map[30][30];
int main()
{
  int n, size = 2, ate = 0; // size 키울 때마다 ate 리셋필요할것
  int dr[4] = {-1, 0, 1, 0};
  int dc[4] = {0, -1, 0, 1};
  int rabbits = 0;
  queue<pair<pair<int, int>, int> > Q;
  scanf("%d", &n);
  for (size_t i = 1; i <= n; i++)
  {
    for (size_t j = 1; j <= n; j++)
    {
      scanf("%d", &map[i][j]);
      if (map[i][j] == 9)
      {
        Q.push(make_pair(make_pair(i, j), 0));
        map[i][j] == 0;
      }
      else if (map[i][j] > 0)
        rabbits++;
    }
  }
  int r, c, time, nr, nc, ntime;
  int chk[30][30] = {0,};
  while (!Q.empty())
  {
    r = Q.front().first.first;
    c = Q.front().first.second;
    time = Q.front().second;
    chk[r][c] = 1;
    printf("%d %d %d\n", r,c,time);
    if (rabbits == 0) break;
    Q.pop();
    for (size_t d = 0; d < 4; d++)
    {
      nr = r + dr[d];
      nc = c + dc[d];
      if (nr <= 0 || nc <= 0 || nr > n || nc > n || chk[nr][nc]) continue;
      ntime = time+1;
      if (map[nr][nc] > 0 && map[nr][nc] < size)
      {
        rabbits--;
        ate++;
        map[nr][nc] = 0;
        if (ate == size)
        {
          size++;
          ate = 0;
        }
        while (!Q.empty()) Q.pop();
        chk[30][30] = {0,};

        Q.push(make_pair(make_pair(nr,nc), ntime)); // 새로운 심바시작점
      }
      else if (map[nr][nc] <= size)
      {
        Q.push(make_pair(make_pair(nr, nc), ntime)); // 그냥 추가.
      }
    }
  }
  printf("%d\n", time);
  return 0;
}